Fix volume control for Frontier Silicon media players

jezcooke opened the pull request:
HA normalises volume controls between 0 -> 1 but Frontier Silicon uses integer steps.  The recent commit 15ed086ed2a1b01ce362cd458c77e22f0fe05f6a improved the issue by multiplying the value by 20; however, not all Frontier Silicon players have the same number of steps (e.g the Roberts R100 that I am using has 40 steps not 20). It now uses fsapi to retrieve the number of steps for the player which should therefore work for all models.

Previously the component did not read back the current volume from the player; therefore if the volume was changed manually or had not been set in HA since the last restart the correct volume would not be shown.  It now implements the volume_level property.  The current volume is queried at the same time as the rest of the media state so will be updated to the correct value within the polling interval.

<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

jezcooke committed:
Fixed volume control for Frontier Silicon media players.

HA normalises volume controls between 0 -> 1 but Frontier Silicon uses integer steps.  The recent commit 15ed086ed2a1b01ce362cd458c77e22f0fe05f6a improved the issue by multiplying the value by 20; however, not all Frontier Silicon players have the same number of steps (e.g the Roberts R100 that I am using has 40 steps not 20). It now uses fsapi to retrieve the number of steps for the player which should therefore work for all models.

Previously the component did not read back the current volume from the player; therefore if the volume was changed manually or had not been set in HA since the last restart the correct volume would not be shown.  It now implements the volume_level property.  The current volume is queried at the same time as the rest of the media state so will be updated to the correct value within the polling interval.

homeassistant commented:
Hi @jezcooke, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#32040).

Once you do that we will be able to review and accept this pull request.

Thanks!

homeassistant commented:
Hello @jezcooke,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`2290d46b52b8c1c4c3d0132642f965994037edae`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;32040&#x2F;commits&#x2F;2290d46b52b8c1c4c3d0132642f965994037edae) - This commit has something that looks like an email address (jcooke@hamamatsu.co.uk). Maybe try linking that to GitHub?.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

jezcooke committed:
Removed unnecessary else which caused pylint test to fail.

jezcooke committed:
Removed whitespace on empty line.

springstan commented on homeassistant/components/frontier_silicon/media_player.py:
```diff
@@ -186,6 +196,9 @@ def media_image_url(self):
             self._source_list = await fs_device.get_mode_list()
 
+        if not self._volume_steps:
+            self._volume_steps = await fs_device.get_volume_steps()
```
Subtract one here so you do not have to do write it multiple times. Why is it subtracted btw?
Would it be useful to cast it to an int as well?
```suggestion
            self._volume_steps = await fs_device.get_volume_steps() - 1
```

jezcooke replied:
Thanks, I've made the suggested changes in the latest commit.

The -1 is because the API seems include 'zero' in the number of steps (e.g. if the range is 0-40 then get_volume_steps returns 41).   Therefore without it a max volume of 1.0 sends it out of range (which causes the API to set the volume back to zero).

springstan commented on homeassistant/components/frontier_silicon/media_player.py:
```diff
@@ -176,6 +178,14 @@ def media_image_url(self):
+        if self._volume is None:
+            return None
+
+        return float(self._volume) / (self._volume_steps - 1)
```
Please create a separate variable `self._volume_level` and move the calculation to `async_update`

jezcooke replied:
Thanks, I've made the suggested changes in the latest commit.

jezcooke committed:
Tweaks from on springstan's suggestions and other fixes

The current volume is now stored normalised (rather than as steps and normalised on request).
The API seems include 'zero' in the number of steps (e.g. if the range is 0-40 then get_volume_steps returns 41).  Now store the steps value with the extra '1' already subtracted (rather than doing it in each calculation).
The volume up/down operations now clamp within the valid range (note that previously going 'up' when already at max volume would reset the volume to zero).
Note: The API reports 'zero' (i.e. turned all the way down, not muted) volume as 'None' rather than '0' hence needing '(volume or 0)' in several places.

springstan commented on homeassistant/components/frontier_silicon/media_player.py:
```diff
@@ -206,6 +216,10 @@ def media_image_url(self):
+
+            volume = await self.fs_device.get_volume()
+
+            self._volume_level = float(volume or 0) / int(self._volume_steps)
```
Isn't `self._volume_steps` always an `int`?
```suggestion
            self._volume_level = float(volume or 0) / self._volume_steps
```

springstan commented on homeassistant/components/frontier_silicon/media_player.py:
```diff
@@ -266,16 +282,18 @@ def is_volume_muted(self):
         volume = await self.fs_device.get_volume()
-        await self.fs_device.set_volume(volume + 1)
+        await self.fs_device.set_volume(
+            min(int(volume or 0) + 1, int(self._volume_steps))
```
Same as above.
```suggestion
            min(int(volume or 0) + 1, self._volume_steps)
```

springstan commented on homeassistant/components/frontier_silicon/media_player.py:
```diff
@@ -266,16 +282,18 @@ def is_volume_muted(self):
     async def async_set_volume_level(self, volume):
         """Set volume command."""
-        await self.fs_device.set_volume(int(volume * 20))
+        await self.fs_device.set_volume(int(volume * int(self._volume_steps)))
```
Same here.
```suggestion
        await self.fs_device.set_volume(int(volume * self._volume_steps))
```

springstan commented on homeassistant/components/frontier_silicon/media_player.py:
```diff
@@ -186,6 +193,9 @@ def media_image_url(self):
             self._source_list = await fs_device.get_mode_list()
 
+        if not self._volume_steps:
+            self._volume_steps = int(await fs_device.get_volume_steps()) - 1
```
Please add a comment on why you have to subtract one from the result.

jezcooke commented on homeassistant/components/frontier_silicon/media_player.py:
```diff
@@ -197,7 +194,7 @@ def volume_level(self):
 
         if not self._volume_steps:
-            self._volume_steps = await fs_device.get_volume_steps()
+            self._volume_steps = int(await fs_device.get_volume_steps()) - 1
```
The API seems to include 'zero' in the number of steps (e.g. if the range is 0-40 then get_volume_steps returns 41), subtract one to get the max volume step.

springstan replied:
Thanks for this information, now your implementation makes sense 👍

springstan replied:
Please add that comment though ✌

jezcooke committed:
Apply suggestions from code review

Co-Authored-By: springstan <46536646+springstan@users.noreply.github.com>

springstan commented on homeassistant/components/frontier_silicon/media_player.py:
```diff
@@ -215,6 +229,8 @@ def media_image_url(self):
             self._media_image_url = None
 
+            self._volume_level = None
+
```
Set volume level to None here as well.
```python
self._volume_level = None
```

jezcooke replied:
Sorry, you've lost me, it's already set to None here?

springstan replied:
I have lost myself 😜 I meant `_volume_steps `

springstan replied:
But looking at it now, it does not make much sense to add it here.

springstan commented on homeassistant/components/frontier_silicon/media_player.py:
```diff
@@ -266,16 +282,18 @@ def is_volume_muted(self):
         volume = await self.fs_device.get_volume()
-        await self.fs_device.set_volume(volume + 1)
+        await self.fs_device.set_volume(
+            min(int(volume or 0) + 1, self._volume_steps)
```
Extract some of it into a variable to make it more readable:
```python
volume = int(volume or 0) + 1
await self.fs_device.set_volume(
            min(volume, self._volume_steps)
)
```

jezcooke replied:
Done

springstan commented on homeassistant/components/frontier_silicon/media_player.py:
```diff
@@ -266,16 +282,18 @@ def is_volume_muted(self):
         """Send volume down command."""
         volume = await self.fs_device.get_volume()
-        await self.fs_device.set_volume(volume - 1)
+        await self.fs_device.set_volume(max(int(volume or 0) - 1, 0))
```
Same here.

jezcooke replied:
Done

jezcooke committed:
Comment to explain why 1 is subtracted from volume_Steps

Also reformatted using black after code review changes.

springstan commented on homeassistant/components/frontier_silicon/media_player.py:
```diff
@@ -266,16 +282,18 @@ def is_volume_muted(self):
     async def async_set_volume_level(self, volume):
         """Set volume command."""
-        await self.fs_device.set_volume(int(volume * 20))
+        await self.fs_device.set_volume(int(volume * self._volume_steps))
```
Same here.

jezcooke replied:
Done

springstan reviewed (commented):
Just some comments to make it more readable.

jezcooke committed:
Split up the set volume functions into steps rather than all in-line as suggested.

springstan commented on homeassistant/components/frontier_silicon/media_player.py:
```diff
@@ -266,16 +284,19 @@ def is_volume_muted(self):
         volume = await self.fs_device.get_volume()
-        await self.fs_device.set_volume(volume + 1)
+        volume = int(volume or 0) + 1
+        await self.fs_device.set_volume(min(volume, self._volume_steps))
```
What do you think about renaming `self._volume_steps` to `self._max_volume`?
Since it contains the value for the highest volume and not the step size in volume (range: 0-40, step size: 1).

jezcooke replied:
I agree, I had already considered renaming it but had left it purely to match the underlying API.  I'll rename it now.

jezcooke committed:
Renamed _volume_steps to _max_volume.

springstan reviewed (approved):
LGTM ✌ thanks for implementing my suggested changes :)

jezcooke committed:
Prevent asnyc_update from failing if we can't get the volume steps

Sometimes the API call can timeout and if async_update fails the media_player becomes unavailable until HA is restarted.

jezcooke commented:
I’ve pushed another commit as I found that occasionally the API to get the volume steps can timeout and if async_update fails the media_player becomes unavailable until HA is restarted.  I thought this was important to fix.

@springstan Incidentally originally I just planned to fix the volume controls  but in the process I found a number of other issues: which I’ve also fixed, namely:

Added the option to specify the name of the device in confirguration.yaml  as sometimes on start-up the component seems to be asked for the name before async_update has got it from the API resulting in the media_player becoming ‘unnamed_device’ thus ends up with a different entity ID breaking automations/Lovelace cards etc. that refer to it (and not looking good of course).  I notice that others have reported the same issue, e.g. here #18556.  I will also need to update the docs for this one.

The player can report itself as ‘off’ when in certain modes (e.g ‘Music player’ or ‘Spofity’ which means HA loses all control (it can’t change input set volume etc as it thinks it’s off).  I’ve changed it so that it reports STATE_IDLE in these cases and only STATE_OFF if it is actually off. This fixes issue #20728.

There seems to be a bug in fsapi where it reports the state as the string “None” (not NoneType) if the first input in the list is selected (‘Internet radio’ in the case of my devices).  I’ve added a workaround to detect this set it to the correct value.

What’s the preferred way of handling this? i.e. should I:

1. Turn this pull request into a general ‘Fix issues with Frontier Silicon media players’ and add everything here.
2. Submit a new pull request for the new issues.
3. Submit individual pull requests for each fix.

springstan commented:
@jezcooke thanks a lot for investigating this integration so thoroughly and planning to fix a lot of issues 👍 

The preferred way of dealing with this is number 3:
> Submit individual pull requests for each fix.

Because this way it is easier to track which changes solve a particular issue :)

jezcooke commented:
@springstan OK, thanks.  New pull requests submitted...

MartinHjelmare reviewed (approved):
Looks good!

MartinHjelmare commented:
Please solve the merge conflict.

jezcooke committed:
Merge branch 'dev' into dev
